### PR TITLE
[SPARK-44242] Improve Max Heap not set check

### PR DESCRIPTION
**Pull request by ashangit**
### What changes were proposed in this pull request?
Currently if any Xmx string is available in the driver java options even if not related to Max Heap setting the job sumission failed
For Ex. this command failed
`./bin/spark-submit --class org.apache.spark.examples.SparkPi --conf "spark.driver.extraJavaOptions=-Dtest=Xmx"  examples/jars/spark-examples_2.12-3.4.1.jar 100`

### Why are the changes needed?
The command should any failed if the -Xmx argument is set not if it is part of another parameter or string

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added a unitary test

### CODE REVIEW

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java [hunk 1]
```diff
@@ -311,11 +311,17 @@ private List<String> buildSparkSubmitCommand(Map<String, String> env)
   }
 
   private void checkJavaOptions(String javaOptions) {
-    if (!isEmpty(javaOptions) && javaOptions.contains("Xmx")) {
-      String msg = String.format("Not allowed to specify max heap(Xmx) memory settings through " +
-        "java options (was %s). Use the corresponding --driver-memory or " +
-        "spark.driver.memory configuration instead.", javaOptions);
-      throw new IllegalArgumentException(msg);
+    System.out.println(javaOptions);
```
Comment (mridulm) [inline launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:314]: Please remove this

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java [hunk 2]
```diff
@@ -311,11 +311,17 @@ private List<String> buildSparkSubmitCommand(Map<String, String> env)
...
+
+    if (!isEmpty(javaOptions)) {
+      for (String javaOption: javaOptions.split(" ")) {
```
Comment (mridulm) [inline launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:317]: This is not robust - you can have space within the value of a property by quoting it.
```
$ scala -Dprop="abc 123"
Welcome to Scala 2.12.14 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_212).
Type in expressions for evaluation. Or try :help.

scala> System.getProperty("prop")
res0: String = abc 123
```

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java (hunk 1 above)
Comment (LuciferYang) [reply launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:314]: +1

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java (hunk 2 above)
Comment (LuciferYang) [reply launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:315]: Maybe `org.apache.spark.launcher.CommandBuilderUtils#parseOptionString` is ok?

Comment (mridulm) [reply launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:315]: Perfect @LuciferYang  !

Comment (ashangit) [reply launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java]: Ah I missed your comment and had reimplemented a kind of parser...
Thks for the pointer I will re-update the code to rely on this one

**Review by mridulm — APPROVED** [review]
Looks good to me.
+CC @LuciferYang

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java [hunk 3]
```diff
@@ -311,11 +311,15 @@ private List<String> buildSparkSubmitCommand(Map<String, String> env)
...
-    if (!isEmpty(javaOptions) && javaOptions.contains("Xmx")) {
-      String msg = String.format("Not allowed to specify max heap(Xmx) memory settings through " +
-        "java options (was %s). Use the corresponding --driver-memory or " +
-        "spark.driver.memory configuration instead.", javaOptions);
-      throw new IllegalArgumentException(msg);
+    if (!isEmpty(javaOptions)) {
+      for (String javaOption: CommandBuilderUtils.parseOptionString(javaOptions)) {
+        if (javaOption.startsWith("-Xmx")) {
+            String msg = String.format("Not allowed to specify max heap(Xmx) memory settings through " +
+                    "java options (was %s). Use the corresponding --driver-memory or " +
```
Comment (LuciferYang) [inline launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:318]: indentation: should be 2 space as before

image

**Comment by LuciferYang** [comment]
@ashangit Java linter check failed, please fix it when you have time

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java (hunk 3 above)
Comment (ashangit) [reply launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java]: Done

File: launcher/src/test/java/org/apache/spark/launcher/SparkSubmitCommandBuilderSuite.java [hunk 4]
```diff
@@ -72,6 +72,36 @@ public void testCliHelpAndNoArg() throws Exception {
       cmd.contains("org.apache.spark.deploy.SparkSubmit"));
   }
 
+  @Test(expected = IllegalArgumentException.class)
+  public void testCheckJavaOptionsThrowException() throws Exception {
+    Map<String, String> env = new HashMap<>();
+    List<String> sparkSubmitArgs = Arrays.asList(
+            parser.MASTER,
```
Comment (LuciferYang) [inline launcher/src/test/java/org/apache/spark/launcher/SparkSubmitCommandBuilderSuite.java:79]: indentation: should be 2 space too

File: launcher/src/test/java/org/apache/spark/launcher/SparkSubmitCommandBuilderSuite.java [hunk 5]
```diff
@@ -72,6 +72,36 @@ public void testCliHelpAndNoArg() throws Exception {
...
+              "-Dprop3='-Xmx -Xmx' -Dprop4='-Xmx \"-Xmx'",
+            SparkLauncher.NO_RESOURCE);
+    buildCommand(sparkSubmitArgs, env);
+  }
+
+  @Test
+  public void testCheckJavaOptions() throws Exception {
+    Map<String, String> env = new HashMap<>();
+    List<String> sparkSubmitArgs = Arrays.asList(
+            parser.MASTER,
```
Comment (LuciferYang) [inline launcher/src/test/java/org/apache/spark/launcher/SparkSubmitCommandBuilderSuite.java:94]: ditto

**Review by LuciferYang — APPROVED** [review]
LGTM, pending ci

**Review by HyukjinKwon — APPROVED** [review]

**Comment by LuciferYang** [comment]
@ashangit Can you rebase the code or re-trigger GA? Some testing tasks failed

**Comment by ashangit** [comment]
Test seems now stuck on pyspak-sql one: https://github.com/ashangit/spark/actions/runs/5451917791/jobs/9918773466#logs without any logs after 4H
Is there a way to relaunch it?

**Comment by LuciferYang** [comment]
You can try canceling the running GA task and then re-trigger the failed one

**Comment by ashangit** [comment]
@mridulm can we merge this PR?

**Comment by mridulm** [comment]
Ah, I thought @LuciferYang might have merged this one :-)

**Comment by mridulm** [comment]
Merged to master.
Thanks for fixing this @ashangit !
Thanks for the reviews @HyukjinKwon, @LuciferYang :-)

**Comment by LuciferYang** [comment]
Sorry, I forgot about that…

**Comment by mridulm** [comment]
Just kidding @LuciferYang :-)

**Pull request closed without merging**